src/Build/Evaluation/Evaluator.cs

```diff
@@ -1393,31 +1411,11 @@ private void EvaluatePropertyElement(ProjectPropertyElement propertyElement)
 
                 _expander.UsedUninitializedProperties.CurrentlyEvaluatingPropertyElementName = null;
 
-                P predecessor = _data.GetProperty(propertyElement.Name);
-
-                P property = _data.SetProperty(propertyElement, evaluatedValue, predecessor);
-
-                if (predecessor != null)
-                {
-                    LogPropertyReassignment(predecessor, property, propertyElement.Location.LocationString);
-                }
-            }
-        }
-
-        private void LogPropertyReassignment(P predecessor, P property, string location)
-        {
-            string newValue = property.EvaluatedValue;
-            string oldValue = predecessor.EvaluatedValue;
-
-            if (newValue != oldValue)
-            {
-                _evaluationLoggingContext.LogComment(
-                    MessageImportance.Low,
-                    "PropertyReassignment",
-                    property.Name,
-                    newValue,
-                    oldValue,
-                    location);
+                P property = _data.SetProperty(
+                    propertyElement,
+                    evaluatedValue,
+                    // If we're tracking property usage, don't first do a get here. The tracker will give false "uninitialized property read" events.
+                    Traits.Instance.LogPropertyTracking ? null : _data.GetProperty(propertyElement.Name));
```

Comment:
You can change the SetProperty method to return the predecessor as an out variable. That way you can treat the predecessor handling in the wrapper and not do any checks here in the Evaluator.

Reply:
This doesn't make sense to me. This is the only instance in the code where this SetProperty overload is called. It actually doesn't care about the predecessor, it just gets it for the method call.

We could entirely remove the predecessor argument from the SetProperty method definition. The only user of it is the Project.Data class. And its SetProperty implementation could just internally call GetProperty to gets the predecessor.

IMO, the above would clean up the code nicely. But I'm not sure if there's something design-wise I'm overlooking.

Reply:
I agree. My initial reasoning was in the terms of "maybe the original intent is to have all accesses visible via the interface", but there's no real actual usage / benefit of that intent, and as you say, there is only one real consumer, Project.Data.SetProperty. Go ahead and remove the predecessor from the interface, and move the logic to Project.Data.SetProperty.

Reply:
Updated next commit.